tools/testbench/topology.c

```diff
@@ -467,16 +537,84 @@ int load_asrc(void *dev, int comp_id, int pipeline_id, int size,
 	return ret;
 }
 
+/* load process dapm widget */
+int load_process(void *dev, int comp_id, int pipeline_id,
+		 struct snd_soc_tplg_dapm_widget *widget)
+{
+	struct sof *sof = (struct sof *)dev;
+	struct sof_ipc_comp_process process = {0};
+	struct sof_ipc_comp_process *process_ipc;
+	struct sof_controls controls = {0};
+	int ipc_size;
+	int size;
+	int ret = 0;
+
+	size = widget->priv.size;
+	ret = tplg_load_process(comp_id, pipeline_id, size, &process, file);
+	if (ret < 0)
+		return ret;
+
+	if (process.comp.type == SOF_COMP_NONE) {
+		fprintf(stderr, "Invalid process comp type SOF_COMP_NONE\n");
+		return -EINVAL;
+	}
+
+	if (tplg_load_controls(&controls, widget->num_kcontrols, file) < 0) {
+		fprintf(stderr, "error: loading controls\n");
+		return -EINVAL;
+	}
+
+	/* Size is process IPC plus private data minus ABI header */
+	ipc_size = sizeof(struct sof_ipc_comp_process);
+	size = 0;
+	if (controls.bytes_ctl) {
+		size = controls.bytes_ctl->priv.size -
+			sizeof(struct sof_abi_hdr);
+		ipc_size += size;
+	}
+	process_ipc = malloc(ipc_size);
+
+	/* Copy header */
+	memcpy(process_ipc, &process, sizeof(struct sof_ipc_comp_process));
+
+	/* Copy configuration data, need to strip ABI header*/
+	memcpy((char *)process_ipc + sizeof(struct sof_ipc_comp_process),
+	       (char *)controls.priv_data + sizeof(struct sof_abi_hdr), size);
+	process_ipc->size = size;
+
+	tplg_free_controls(&controls);
+
+	/* load process component */
+	register_comp(process_ipc->comp.type);
+
+	/* Instantiate */
+	if (ipc_comp_new(sof->ipc, (struct sof_ipc_comp *)process_ipc) < 0) {
+		fprintf(stderr, "error: new process comp\n");
+		return -EINVAL;
+	}
+
+	return ret;
```

Comment:
This looks like we are starting to duplicate a lot of component loading code. Can we make this all generic.

Reply:
I simplified the controls load / ignore from previous version but otherwise I didn't see similar flows. There's always some component specifics.